toolchain/parse/context.cpp

```diff
@@ -427,6 +430,46 @@ auto Context::EmitExpectedDeclSemiOrDefinition(Lex::TokenKind expected_kind)
   emitter().Emit(*position(), ExpectedDeclSemiOrDefinition, expected_kind);
 }
 
+// Returns whether we are currently parsing in a context in which methods can be
+// declared, such as a class or interface.
+static auto ParsingInMethodContext(Context& context) -> bool {
+  auto& stack = context.state_stack();
+  if (stack.size() < 2 || stack.back().state != State::DeclScopeLoop) {
+    return false;
+  }
+  auto state = stack[stack.size() - 2].state;
+  return state == State::DeclDefinitionFinishAsClass ||
+         state == State::DeclDefinitionFinishAsImpl ||
+         state == State::DeclDefinitionFinishAsInterface ||
+         state == State::DeclDefinitionFinishAsNamedConstraint;
+}
+
+auto Context::AddFunctionDefinitionStart(Lex::TokenIndex token,
+                                         int subtree_start, bool has_error)
+    -> void {
+  if (ParsingInMethodContext(*this)) {
+    enclosing_inline_method_stack_.push_back(tree_->inline_methods_.Add(
```

Comment:
I wonder... beyond just having the parse tree track members to out-of-line, could the parse tree produce an ordering of parse nodes as they should be analyzed, e.g. as the chunks? Not blocking this PR; just generally wondering if you think that might be viable/worthwhile.

Reply:
Hm. Yes, I like the idea, and I think that could work well. I suppose this would amount to the parser producing a list of actions:

- process chunk from NodeId A to NodeId B
- suspend function
- resume function
- suspend scope
- resume scope
- leave scope

and the check side of things going back to a simple in-order execution of the actions given to it by the parser.

But yeah, I'd prefer to land this PR first. Most of the pieces here should be reusable in that approach.